### DISCUSSION TITLE
tutorials/inputs/controllers_gamepads_joysticks

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#14] [User-contributed notes]
# tutorials/inputs/controllers_gamepads_joysticks

Godot supports hundreds of controller models thanks to the community-sourced SDL game controller database. Controllers are supported on Windows, macOS, Linux, Android, iOS, and HTML5. Note that mor...

https://docs.godotengine.org/en/latest/tutorials/inputs/controllers_gamepads_joysticks.html

**Comment by NancokPS2** [comment]
It is a known issue that Switch controllers don't work, they seem to "spam" inputs constantly.   
Please refer to this issue: [#81191](https://github.com/godotengine/godot/issues/81191)

**Comment by Odra08** [comment]
it should be better if there could be added a documentation for usb serial/degugging, for them that are using arduino or raspberry0 serial comunication, may also other devices like that.
you wandering what a arduino or a raspberry is?
look here:
https://www.arduino.cc
https://www.raspberrypi.com

**Reply by Calinou** [reply]
There's a third-party GDExtension adding serial support: https://github.com/matrixant/serial_port
According to its issues page, it doesn't seem to work on 4.3 though.

Other than that, feature proposals should be made on the [godot-proposals](https://github.com/godotengine/godot-proposals) repository.

**Comment by norapatches** [comment]
"If you want controller buttons to send echo events, you will have to generate InputEvent objects by code and parse them using Input.parse_input_event() at regular intervals. This can be accomplished with the help of a Timer node."

I would like the d-pad up and down to move my camera up/down in a 2D platformer but only if I'm holding it for a certain amount of time. Now it only works with keyboard WASD but not with a controller (PS4) because holding the d-pad down doesn't echo and hence my Input.is_action_pressed() logic doesn't get executed.
How to make it universal so it generates press event only on gamepad and how to use Input.parse_event(event) at all?

```gdscript
func handle_input(event: InputEvent, action_map: Dictionary) -> void:
	for action in action_map:
		# PRESS
		if Input.is_action_just_pressed(action) and (event is InputEventKey or event is InputEventJoypadButton or event is InputEventJoypadMotion):
			action_map[action]['press'].call()
			self.action_down_times[action] = Time.get_ticks_msec()
		# HOLD
		elif event.is_action_pressed(action, true):
			action_map[action]['hold'].call(Time.get_ticks_msec() - action_down_times[action])
		# RELEASE
		elif Input.is_action_just_released(action) and (event is InputEventKey or event is InputEventJoypadButton or event is InputEventJoypadMotion):
			if action in action_up_times:
				release_delta = Time.get_ticks_msec() - action_up_times[action]
			action_map[action]['release'].call(double_action_threshold > release_delta and release_delta != 0)
			self.action_up_times[action] = Time.get_ticks_msec()
```

**Comment by InfGame** [comment]
Is there a way to differentiate Left joystick and right joystick inputs?

**Reply by NancokPS2** [reply]
Yes, using the `axis` property. But this is not the place to make support questions.

**Comment by PhoenixAstra1** [comment]
I made an echo script for my DPAD and controller to work and struggle to get them both working. I can't get them to work with my scroll bar. It works with a regular press but not with echo no matter what I do, it lags behind even if I try to force scroll

Anyone deal with echo dpad/controller repeats and with a scrollcontainer?

**Comment by felipebottega** [comment]
"Unlike keyboards and mice" lol

**Reply by Calinou** [reply]
Analog keyboards exist[^1], but they send their analog input through controller emulation (typically Xbox 360 controller, as a lot of games support it). Operating systems don't have APIs to get the analog input on a *key*, so this is still technically valid.

[^1]: Yours truly has a Wooting two HE around :wink:

**Reply by felipebottega** [reply]
@Calinou I was referring to the term "mice." It sounded funny. Shouldn't it be "mouse"?

**Reply by Calinou** [reply]
> Shouldn't it be "mouse"?

The plural of "mouse" is "mice" :slightly_smiling_face:
